### PR TITLE
pallet-evm: add support for tuple-based precompile declarations

### PR DESCRIPTION
**Pull request by sorpaas**
This adds support for tuple-based precompile declarations, to simplify writing them. If `type Precompiles` is a tulpe, then the first item will be callable at `address(1)`, the second one at `address(2)`, and so forth.

### CODE REVIEW

File: frame/evm/src/precompiles.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
+use sp_std::vec::Vec;
```
Comment (kianenigma) [inline frame/evm/src/precompiles.rs:1]: license missing.

File: frame/evm/src/precompiles.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	/// Try to execute the precompile. Calculate the amount of gas needed with given `input` and
+	/// `target_gas`. Return `Ok(status, output, gas_used)` if the execution is
+	/// successful. Otherwise return `Err(_)`.
+	fn execute(
+		input: &[u8],
+		target_gas: Option<usize>,
+	) -> core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError>;
+}
+
+macro_rules! impl_precompile_tuple {
```
Comment (kianenigma) [inline frame/evm/src/precompiles.rs:40]: considered using impl_trait_for_tuples?

Comment (bkchr) [reply frame/evm/src/precompiles.rs:40]: Yes, please use `impl_trait_for_tuples` for this!

Comment (bkchr) [reply frame/evm/src/precompiles.rs:40]: Just search the code base for this.

Comment (sorpaas) [reply frame/evm/src/precompiles.rs]: Fixed!

**Comment by sorpaas** [comment]
polkadot-companion-build fail unrelated to this PR, and will be fixed when https://github.com/paritytech/polkadot/pull/1394 is merged.

File: frame/evm/src/precompiles.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	/// Try to execute the precompile. Calculate the amount of gas needed with given `input` and
+	/// `target_gas`. Return `Ok(status, output, gas_used)` if the execution is
+	/// successful. Otherwise return `Err(_)`.
+	fn execute(
+		input: &[u8],
+		target_gas: Option<usize>,
+	) -> core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError>;
+}
+
+#[impl_for_tuples(1, 16)]
```
Comment (bkchr) [inline frame/evm/src/precompiles.rs:58]: Putting here just `impl_for_tuples(16)` should enable you to drop the custom impl for `()` above?

**Pull request merged**